doc: best practices within multi-tenancy solutions (#5353)
operator-framework/operator-sdk: website/content/en/docs/best-practices/managing-resources.md

- `Memory`: largely based on the number of primary resources that exist (multiplied by some factor based on the number of operand resources they need to watch as a result) via informer caches.

And then there is a concern that one Pod or Container could monopolize all available resources and Cluster admins must 
consider the effects that one Pod or Container may have on other components. 

In an effort to prevent a container from consuming all the resources on a cluster or affecting other workloads
from being scheduled, many production clusters will define [ResourceQuota][k8s-resources-quotas] configurations.

The [ResourceQuota][k8s-resources-quotas] configuration also applies to tenant workloads that are managed by
your Operator. Cluster administrators will typically set a [ResourceQuota][k8s-resources-quotas] for each
tenant's namespace as part of the onboarding. If a [LimitRange][k8s-limit-range] with default values
has **not** been created in each namespace and your Operator creates Containers inside the tenant
namespace without specifying at least resource requests for CPU and Memory of its Pods then,

> the system or quota may reject Pod creation with errors such as, for example: `[maximum cpu usage per Pod is 100m.  No limit is specified, maximum memory usage per Pod is 512Mi.  No limit is specified]`. 

Reviewer: you're recommending people need to set `requests` but this error is complaining about missing `limits`

Author: Because the rule was for the max value. 
Removed the e.g. We do not need that.

> the system or quota may reject Pod creation with errors such as, for example: `[maximum cpu usage per Pod is 100m.  No limit is specified, maximum memory usage per Pod is 512Mi.  No limit is specified]`. 

Author: ```suggestion
the system or quota may reject Pod creation.
```

Check the following statements obtained from K8s docs:

> - > If a LimitRange is activated in a namespace for computing resources like CPU and memory, users must specify requests or limits for those values. Otherwise, the system may reject Pod creation. (  [Reference][k8s-limit-range] ).

Author: ```suggestion
- > If a LimitRange is activated in a namespace for computing resources like CPU and memory, users must specify requests or limits for those values. Otherwise, the system may reject Pod creation. ([Reference][k8s-limit-range]).
```

> - > If quota is enabled in a namespace for compute resources like cpu and memory, users must specify requests or limits for those values; otherwise, the quota system may reject pod creation. ( [Reference][k8s-resources-quotas] ).

Author: ```suggestion
- > If quota is enabled in a namespace for compute resources like cpu and memory, users must specify requests or limits for those values; otherwise, the quota system may reject pod creation. ([Reference][k8s-resources-quotas]).
```

---
title: "Managing Resources"
linkTitle: "Managing Resources"
weight: 5
description: This guide describes good practices concepts to Managing Resources.
---

## Managing Resources

Note that pretty much all controllers consume:

- `CPU`: largely based on the number of reconciliations they perform, which are generally related to event activity for resources they're watching.
- `Memory`: largely based on the number of primary resources that exist (multiplied by some factor based on the number of operand resources they need to watch as a result) via informer caches.

> And then, there is a concern that one Pod could monopolize all available resources and Cluster admins must 

Reviewer: ```suggestion
And then there is a concern that one Pod could monopolize all available resources and Cluster admins must 
```

> consider the effects that one container may have on others components. 

Reviewer: ```suggestion
consider the effects that one container may have on other components. 
```

> consider the effects that one container may have on others components. 

Author: ```suggestion
consider the effects that one Pod may have on others components. 
```

In an effort to prevent a container from consuming all the resources on a cluster or affecting other workloads
from being scheduled, many production clusters will define [ResourceQuota][k8s-resources-quotas] configurations.

The [ResourceQuota][k8s-resources-quotas] configuration also applies to tenant workloads that are managed by
your Operator. Cluster administrators will typically set a [ResourceQuota][k8s-resources-quotas] for each
tenant's namespace as part of the onboarding. If a [LimitRange][k8s-limit-range] with default values
has **not** been created in each namespace and your Operator creates Pods/Deployments inside the tenant
namespace without specifying at least resource requests for CPU and Memory of its Pods then,

> those Pods may fail to schedule and tenants may not be able to use your Operator.

Reviewer: > If a LimitRange is activated in a namespace for computing resources like CPU and memory, users must specify requests or limits for those values. Otherwise, the system may reject Pod creation.
> If quota is enabled in a namespace for compute resources like cpu and memory, users must specify requests or limits for those values; otherwise, the quota system may reject pod creation. 

rejecting pod creation is a little different than failure to schedule (same end result, but looks different to the user).  So we should properly describe what will actually happen.

Author: Done.

In an effort to support clusters with this configuration, to ensure safe operations and avoid negatively

> impacting other workloads: Operators should **always** include reasonable memory and CPU resource requests. 

Reviewer: ```suggestion
impacting other workloads: Operators should **always** include reasonable memory and CPU resource requests for their own deployment as well as for operands they deploy. 
```

better?

Therefore, your Operator should **always** apply at least resource requests for `CPU` and
memory for `Pods/Deployments` that it creates as part of the reconciliation logic. Ideally your
Operator also applies memory limits to those `Pods/Deployments`. You may also consider CPU limits.

Resource requests and limits for the Operator Deployment can be defined by modifying the`config/manager/manager.yaml`
as shown below:

```yml
  ...
  # TODO(user): Configure the resources accordingly based on the project requirements.
  # More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
  resources:
    requests:
      cpu: 10m
      memory: 64Mi
    ...
```

> ### But, how to figure out the ideal values

Reviewer: ```suggestion
### How to compute default values
```

**IMPORTANT:** A single configuration that fits all scenarios is not possible. 
In this way, Operators authors **MUST** to ensure that Cluster Admins and its users can change the 
resource requests/limits of the Operator/manager, and of its Operands.

However, you are able to benchmark your Operator by [Monitoring the resource usage][k8s-metrics] 
to ensure good and reasonable values for the general cases. Kubebuilder and SDK provide some 
[metrics][kb-metrics] which can help you with.

**NOTE** Also, be aware that if the project was generated by Kubebuilder or SDK scaffold
then some values for the `Operator/manager`(See `config/manager/manager.yaml`)
are populated by default to get you started, however, you ought to optimize them based 
on your own tests and the specific needs of your operator.

> ### When integrating with OLM, how to allow change the Operator/manager resources values

Reviewer: ```suggestion
### How to change the Operator/manager resources values when under OLM management
```

If your operator is managed by OLM, administrators or users can configure your operator's resource requests and limits
via the [subscription][olm-subscriptions]. 

### General guidelines

Following are some general recommendations to manage the resources:

> **NOTE** If your operator is managed by OLM, administrators or users can configure your operator's resource requests and limits
> via the [subscription][olm-subscriptions]). Also, be aware that if the project was generated by Kubebuilder or SDK scaffold
> then some values for the `Operator/manager`(See `config/manager/manager.yaml`)
> are populated by default to get you started, however, you ought to optimize them based on your own tests and the specific needs of your operator.

Author: ```suggestion
```

  and if the `Pod` starts hitting the CPU limits, Kubernetes uses kernel to starts throttling the container.
  That means the CPU will be artificially restricted, giving a potentially worse performance only.

You might want to check the [Troubleshooting][k8s-troubleshooting] section in the Kubernetes documents to better
understand how to debug these scenarios.

#### Limits are specified but not requests

_What happens when the resource limits are defined but not the requests?_

If you specify a CPU or Memory limit for a Container but do not specify a request, 
Kubernetes automatically assigns a CPU or Memory request that matches the limit. 
In this way, you will be requesting **always** the limit and will be allocating 
more resources than required. (**NOT RECOMMENDED**)

#### Values are too big

Memory and CPU requests and limits are associated with Containers, but be aware that the Memory and CPU 
requests and limits of a Pod are the sum of its specific computing types for all the containers in the Pod. 

If you define that your Pods should have Memory or CPU request too big then, you might not only be allocating and 

> blocking the usage of more than you ought unnecessarily. Also, your Operator consumers might be unable to install them 

Reviewer: saying consumers won't be able to install them is a bit confusing.

yes technically the OLM install may fail, because the operator deployment will never be able to get scheduled (if no node has sufficient resources to meet the request), but i'd say it's more that consumers won't be able to "deploy" the operator than that they won't be able to "install" it.  The operator installs, the deployment of the operator fails.

Author: Tried to address it. Please, feel free to re-check.
